adagrams/game.py

```diff
@@ -1,11 +1,139 @@
+            hand.append(random_letter)
 
+    return hand
+
+#helper function to convert a string or list into a dictionary with the amount of times that element appears
+#as it value
+def word_letter_dictionaries(data_to_convert):
+    dict_convert = {}
+
+    for elem in data_to_convert:
+        if elem in dict_convert:
+            dict_convert[elem] +=1
+        else:
+            dict_convert[elem]= 1
+    return dict_convert
+    
 def uses_available_letters(word, letter_bank):
-    pass
+    #valid variable to check whether the letter is in letter_bank or not
+    valid = False
+    word_upper = word.upper()
+    #dict_word and dict_letter_bank store a dictionary with each letter as key and the amount of that that letter
+    #appears as value
+    dict_word = word_letter_dictionaries(word_upper)
+    dict_letter_bank = word_letter_dictionaries(letter_bank)
+    #for loop to loop through dict_word  and check whether the letters are in dict_letter_bank
+    #or not and in case the amount of that letter exceeds the amount in dict_letter_bank it returns False
+    for elem in dict_word:
+        if elem not in letter_bank or dict_word[elem] > dict_letter_bank[elem]:
+            valid = False
+            break
+        else:
+            valid = True
+
+    return valid
 
 def score_word(word):
-    pass
+    #create dict_score to store letters that shares the same point value
+    dict_score = {1: ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
+                  2: ["D", "G"],
+                  3: ["B", "C", "M", "P"],
+                  4: ["F", "H", "V", "W", "Y"],
+                  5: ["K"],
+                  8: ["J", "X"],
+                  10: ["Q", "Z"]}
+    #score variable to store the points earned
+    result = 0
+    #word_upper_case variable converts word variable elements into uppercase letters
+    word_upper = word.upper()
+    #for to loop through the elements in word_upper to give the right points per letter
+    for letter in word_upper:
+        for score, letters in dict_score.items():
+            if letter in letters:
+                result += score
+            
+    #if word has 7, 8, 9 or 10 letters the word gets 8 additional points
+    if len(word) >= 7 and len(word)<= 10:
+        result += 8
+        
+    return result
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    #tuple to store the word with the high score
+    high_score_tuple = ()
+    #dictionary to store all the words with their score
+    score_dict = {}
+    score = 0
+    #append every word with its score in the dictionary
+    for word in word_list:
+        score = score_word(word)
+        score_dict[word]= score
+    #get the word with the maximun score in score_dict
+    value_high_score = max(score_dict.values())
+    #append the words that share the same store into list_high_word
+    list_high_word = [word for word, score in score_dict.items() if score == value_high_score]    
```

Comment:
Great way to filter down to a list of only words that were tied for the high score. We now know that everything in the list is a winning candidate and that we don't need to think about their score any longer, making the remaining checks simpler.